include/jemalloc/internal/tsd.h

```diff
@@ -74,7 +75,7 @@ typedef void (*test_callback_t)(int *);
     O(iarena,			arena_t *,		arena_t *)	\
     O(arena,			arena_t *,		arena_t *)	\
     O(arenas_tdata,		arena_tdata_t *,	arena_tdata_t *)\
-    O(binshard,			unsigned,		unsigned)	\
+    O(binshards,		tsd_binshards_t,	tsd_binshards_t)\
```

Comment:
This should live after the tcache, right? (Or at least, in the tcache bonus area). We don't touch it on any fast paths.

I mean, for the record I still think trying to micro-optimize to that extent is a penny-wise pound-foolish thing that costs us more in dev velocity than it gains us in cycles, but if that's what we're going to decide we should decide it explicitly, and stop trying to do it elsewhere, too.

Reply:
Oh I was mainly thinking we will want to use this to do per thread, per size binding decisions, e.g. when we have enough contentions, switch this thread to another shard for a specific size class. Agree that the few cycles it saves isn't really a big deal, but combined with the flexibility it gives, might as well have it from the beginning.

The place for this is pretty awkward, it's not exactly tiny (36B), but coupling it into tcache doesn't feel right too. I can imagine for manual arenas people will want to bypass tcache with the help of the sharded bins to avoid too much contention. Put it after tcache may touch an extra page; also likely introduce TLB miss for a semi-fast path (tcache fill).

I slightly prefer to leave it as it, but also fine switching back to a single field and reduce to the range on the fly.

Reply:
I mean, I agree with what you're saying, but that's not the decision procedure we've done in the past; if we think cleanliness is more important than struct layout micro-optimization, we should move (e.g.) tsd_link and in_hook here too, and not bother with the large vs. small distinction. That was something you insisted on, so if you've changed your mind on how important it is we ought to clean up the code, too.

Reply:
This one is quite on the boardline. On one hand it's required on every tcache fill / flush, so I didn't want it to be at the tail (also combined with the possibility of touching another page); on the other hand it's not really fast path and takes half a cacheline here. I agree that the tcache bonus area is a good fit; the reason pushed me away from that was the recent struggling the rocksDB guys had "wanting a single manual arena; wanting manual tcache to improve scalability; sadly manual tcache increases RSS (w/ >1k threads)". So I made this independent from tcache.

But I think we should try a bit harder here; how about I do a follow up patch to move this into tcache; and then for the allocations w/o tcache we make on-the-fly sharding decisions?